enhancements/insights/container-scanner.md

```diff
@@ -0,0 +1,532 @@
+  * __`Name`__ - Name of the runtime used to run the application in the container
+    * Derived value (set by the container scanner)
+    * Optional - based on the capabilities of the container scanner to detect such runtimes
+    * Value is a hash of the detected runtimes `Quarkus, Spring Boot, Apache Tomcat, WildFly, JBoss EAP, RH-SSO, Kafka <x.y>, PostgreSQL` (exhaustive list based on the capabilities of the container scanner, might be enhanced with additional development)
+  * __`Version`__ - The version of the runtime used to run the application
+    * Raw values (read from files in the containers)
+    * Optional - based on the capabilities of the container scanner to detect the runtime and the availability of its version from files in the container
+    * Value is a hash of the detected runtime version: `2.10.3.Final, 7.4.7.GA, 2.7.1, 3.6.0.redhat-00005,...` (list not exhaustive)
+
+It is not planned to add more fields to the `workloadRuntimeInfoContainer` type. However, the container scanner can be enhanced to detect more runtimes without impacting the integration in the Insights Operator.
+
+##### Examples
+
+```json
+"runtimeInfo": {
+  "os": "yqx1WBefoIAq",
+  "osVersion": "AQN-qh_OyRUA",
+  "kind": "wbpgzhNYZQOi",
+  "kindVersion": "cAsoUDy8JFNV",
+  "kindImplementer": "DzP5-GS6tDOQ",
+  "runtimes": [
+    {
+      "name": "zBVuVhrrC_vC",
+      "version": "lyAE7Oh63WPU"
+    }
+  ]
+}
+```
+
+As the values are derived by the container scanner or correspond to software versions, it is possible to create a dictionary of the hashes and retrieve the unobfuscated values. The example above corresponds to:
+
+```json
+"runtimeInfo": {
+  "os": "rhel",
+  "osVersion": "8.9",
+  "kind": "Java",
+  "kindVersion": "17.0.7",
+  "kindImplementer": "Red Hat, Inc.",
+  "runtimes": [
+    {
+      "name": "Quarkus",
+      "version": "2.13.8.Final-redhat-00004"
+    }
+  ]
+}
+```
+
+### Workflow Description
+
+In its existing implementation, the OpenShift Insights Operator periodically gathers workload data every 12 hours.
+
+The Operator iterates over pods in the cluster (up to a limit of 8000) and, for each pod's containers, captures their Image ID and entry points.
+
+When the operator starts gathering workload information, it would first gather runtime information for all the containers running on the cluster's worker nodes and then
+populate the existing gathered info with these additional runtime information.
+
+![container scannner workflow](./container-scanner-diagram-1.png "Container Scanner Worklow")
+
+This is not a user-facing feature. It is achieved within the OpenShift Insights Operator gathering workflow.
+The user is able to examine the gathered data by unarchiving the archive generated by the operator.
+The DaemonSet for the container scanner is deployed on-demand when the operator gathers workload data and is not accessible outside of the `openshift-insights` namespace.
```

Comment:
I am still thinking about the integration approach here. What advantages do you see in repeatedly deploying DaemonSet please? Maybe it would be easier to have it deployed once during the cluster installation and initiate/trigger the collection only when needed. 
It would save the waiting time for the deployment to be ready. It would eliminate the pruning/undeployment. The only disadvantage I can see is that it would consume some resources all the time.

Reply:
We could bind the daemonset lifecycle to the operator. I see some pros with the proposed approach:

* As you mention, we don't consume resources for something that is actually running every 12 hours. There will be memory consumption when we collect the data but all this memory can be discarded when the gather is done instead of continue to allocate it for the rest of the daemonset pod's lifetime.
* We only care about the cluster topology when the gather operation is performed. Worker nodes can be up and down in the mean time. If the dameonset was always up, we would have to make sure that is properly replicated on every node and that they are all responding. With the current approach, we deploy on the worker nodes right before  the gather so their previous states has no impact.
* The DaemonSet has high permissions and I'd be more comfortable to reduce its availability (we need to make sure that it's not a trampoline pod whether it's running a few minutes every 12 hours or 24/7 so this is not a important point)
* The deployment time should be very small (the image is preloaded in the cluster and it does nothing at boot time)

I defer to you on the integration approach. We could deploy once as you suggest and keep the proposed approach as an optimization down the road if needs be.

Reply:
Thanks. I don't know what the best integration approach would be, because it seems that there will always be some trade-offs. 
I wonder if using batch API would make sense, because as you mentioned, we are not interested in the previous state and it's basically one-off action. One drawback I can see is reading the data, that is not possible from a completed Pod so the job would probably need to run/wait for the operator to consume the data (and then kill the job?)

Reply:
I'm resolving this conversation as the RFE was updated to deploy the insights-operator-extractor when the operator is installed.